staging/src/k8s.io/component-base/metrics/collector.go

```diff
@@ -0,0 +1,155 @@
+// metric registration process, it's especially intend to be used in scenario of custom collector.
+//
+// Any custom collector only need to implements the 'CustomCollect' method,
+// and the other methods are implemented by 'BaseCustomCollector'.
+type CustomCollector interface {
+	prometheus.Collector
+
+	// CustomDescribe sends the super-set of all possible metrics.Desc collected
+	// by this CustomCollector to the provided channel.
+	CustomDescribe(chan<- *Desc)
+
+	// CustomCollect sends each collected metrics.Metric via the provide channel.
+	CustomCollect(chan<- Metric)
+
+	// Create will initialize all Desc that has been added by Add(). It will be called when registry.
+	Create(version *semver.Version) bool
+
+	// Init accepts custom collector itself it intends to be called when registry.
+	Init(self CustomCollector)
+}
+
+// BaseCustomCollector which implements almost all of the methods defined by CustomCollector
+// is a convenient assistant for custom collectors.
+// It is recommend that inherit BaseCustomCollector when implementing custom collectors.
+type BaseCustomCollector struct {
+	descriptors []*Desc // stores all Desc collected from CustomDescribe().
+	registrable []*Desc // stores registrable Desc(not be hidden), is a subset of descriptors.
+	hidden      []*Desc // stores hidden Desc
+	self        CustomCollector
+}
+
+// CustomDescribe sends all descriptors to the provided channel.
+// Every custom collector should over-write this method.
+func (cc *BaseCustomCollector) CustomDescribe(ch chan<- *Desc) {
+	panic(fmt.Errorf("custom collector should over-write CustomDescribe method"))
+}
+
+// Describe sends all descriptors to the provided channel.
+// It intend to be called by prometheus registry.
+func (cc *BaseCustomCollector) Describe(ch chan<- *prometheus.Desc) {
+	for _, d := range cc.descriptors {
+		ch <- d.ToPrometheusDesc()
+	}
+}
+
+// CustomCollect sends all metrics to the provided channel.
+// Every custom collector should over-write this method.
+func (cc *BaseCustomCollector) CustomCollect(ch chan<- Metric) {
+	panic(fmt.Errorf("custom collector should over-write CustomCollect method"))
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+func (cc *BaseCustomCollector) Collect(ch chan<- prometheus.Metric) {
+	mch := make(chan Metric)
+
+	go func() {
+		cc.self.CustomCollect(mch)
+		close(mch)
+	}()
+
+	for m := range mch {
+		// Hidden metrics should be ignored.
+		// TODO(RainbowMango): There is no convenient method to identify if the metrics should be ignored.
+		// 	Use a temporary solution here. (try to search in hidden list)
+		if strings.Contains(m.Desc().String(), hiddenFlag) {
+			continue
+		}
+
+		ch <- prometheus.Metric(m)
+	}
+}
+
+func (cc *BaseCustomCollector) add(d *Desc) {
+	cc.descriptors = append(cc.descriptors, d)
+}
+
+// Init intends to be called by registry.
+func (cc *BaseCustomCollector) Init(self CustomCollector) {
```

Comment:
Why is this public? Is there a reason you don't invoke this in the Create?

Reply:
Removed `Init()` from the interface and changed to private, invoked by `Create()`.